components/catalogs/back/src/Infrastructure/Service/ProductMapper.php

```diff
@@ -0,0 +1,126 @@
+<?php
+
+declare(strict_types=1);
+
+
+namespace Akeneo\Catalogs\Infrastructure\Service;
+
+use Akeneo\Catalogs\Application\Persistence\Catalog\Product\GetRawProductsQueryInterface;
+use Akeneo\Catalogs\Application\Service\ProductMapperInterface;
+use Akeneo\Catalogs\Application\Storage\CatalogsMappingStorageInterface;
+use Akeneo\Catalogs\Domain\Catalog;
+use Akeneo\Catalogs\ServiceAPI\Exception\ProductSchemaMappingNotFoundException as ServiceApiProductSchemaMappingNotFoundException;
+use Akeneo\Catalogs\ServiceAPI\Query\GetMappedProductsQuery;
+
+/**
+ * @copyright 2022 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ *
+ * @phpstan-import-type MappedProduct from GetMappedProductsQuery
+ * @phpstan-import-type RawProduct from GetRawProductsQueryInterface
+ * @phpstan-import-type ProductMapping from Catalog
+ */
+class ProductMapper implements ProductMapperInterface
+{
+    public function __construct(
+        private readonly CatalogsMappingStorageInterface $catalogsMappingStorage,
+    ) {
+    }
+
+    public function getMappedProducts(array $products, Catalog $catalog): array
+    {
+        $productMappingSchema = $this->getProductMappingSchema($catalog->getId());
+        $productMapping = $catalog->getProductMapping();
+
+        return \array_map(
+            /** @param RawProduct $product */
+            fn (array $product): array => $this->mapProduct($product, $productMappingSchema, $productMapping),
+            $products
+        );
+    }
+
+    public function getMappedProduct(array $product, Catalog $catalog): array
+    {
+        $productMappingSchema = $this->getProductMappingSchema($catalog->getId());
+        $productMapping = $catalog->getProductMapping();
+
+        return $this->mapProduct($product, $productMappingSchema, $productMapping);
+    }
```

Comment:
```suggestion
    public function getMappedProduct(array $product, Catalog $catalog): array
    {
        $productMappingSchema = $this->getProductMappingSchema($catalog->getId());
        $productMapping = $catalog->getProductMapping();

        return $this->mapProduct($product, $productMappingSchema, $productMapping);
    }
```

With the previous suggestion, I was telling you that we should loop externally.
It raise one problem, the mapping schema will be loaded several times.

To prevent this, we have 2 options:
- we have a LRU cache in the class (stateful)
- we change the signature of the mapper to take the schema & the mapping as arguments instead of the Catalog.